File: ios/MunchiSunmiPrinter.m
#if __has_include(<React/RCTBridgeModule.h>)
  #import <React/RCTBridgeModule.h>
#elif __has_include("RCTBridgeModule.h")
  #import "RCTBridgeModule.h"
#else
  #import "React/RCTBridgeModule.h"
#endif

@interface RCT_EXTERN_MODULE(MunchiSunmiPrinter, NSObject)

RCT_EXTERN_METHOD(initialize:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(prepare:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(getDeviceInfo:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(getPrinterStatus:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(print:(NSDictionary *)job resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(printText:(NSString *)text options:(NSDictionary *)options resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(cutPaper:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(lineWrap:(nonnull NSNumber *)lines resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)

@end

File: ios/MunchiSunmiPrinter.swift
import Foundation

#if canImport(React)
import React
#elseif canImport(React_Core)
import React_Core
#endif

@objc(MunchiSunmiPrinter)
class MunchiSunmiPrinter: NSObject {
  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }

  @objc(initialize:rejecter:)
  func initialize(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    reject("IOS_NOT_SUPPORTED", "SUNMI printer is not supported on iOS", nil)
  }

  @objc(prepare:rejecter:)
  func prepare(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    reject("IOS_NOT_SUPPORTED", "SUNMI printer is not supported on iOS", nil)
  }

  @objc(print:resolver:rejecter:)
  func print(
    _ job: NSDictionary,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    reject("IOS_NOT_SUPPORTED", "SUNMI printer is not supported on iOS", nil)
  }

  @objc(getDeviceInfo:rejecter:)
  func getDeviceInfo(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    resolve([
      "manufacturer": "apple",
      "brand": "apple",
      "model": "ios",
      "sdkInt": 0
    ])
  }

  @objc(getPrinterStatus:rejecter:)
  func getPrinterStatus(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    resolve([
      "online": false,
      "paperEmpty": false,
      "coverOpen": false,
      "errorCode": "UNSUPPORTED_PLATFORM",
      "errorMessage": "The SUNMI printer bridge is not supported on this platform.",
      "nativeStatusCode": NSNull()
    ])
  }

  @objc(printText:options:resolver:rejecter:)
  func printText(
    _ text: String,
    options: NSDictionary?,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    reject("IOS_NOT_SUPPORTED", "SUNMI printer is not supported on iOS", nil)
  }

  @objc(cutPaper:rejecter:)
  func cutPaper(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    reject("IOS_NOT_SUPPORTED", "SUNMI printer is not supported on iOS", nil)
  }

  @objc(lineWrap:resolver:rejecter:)
  func lineWrap(
    _ lines: NSNumber,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    reject("IOS_NOT_SUPPORTED", "SUNMI printer is not supported on iOS", nil)
  }
}
